proposed/2021/PackageNamespaces.md

```diff
@@ -0,0 +1,568 @@
+A user may also pin a specific package id instead of the complete namespace.
+Specific ids take precedence over the namespaces.
+
+**Example:**
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <!-- Add a namespace for Microsoft.* or NuGet.* under the nuget.org source. -->
+    <packageSource name="nuget.org">
+        <namespace id="Microsoft.*" />
+        <namespace id="NuGet.*" />
+    </packageSource>
+
+    <!-- Add a namespace for Contoso.* under the contoso source. -->
+    <packageSource name="contoso">
+        <namespace id="Contoso.*" />
+        <!-- Add a specific id to be download from  the contoso source. -->
+        <namespace id="Special.Package.With.An.Imperfect.Id" />
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+By default, namespaces will allow a package to match multiple namespaces and download with no precedence. In some cases, one might want to avoid ambiguous cases. That can be done by adding a `strict` flag.
+
+```xml
+<config>
+    <add key="namespaceMode" value="strict" />
+</config>
+```
+
+The strict flag provides a source pinning behavior by nature in which one namespace is allowed per package source. Entering a strict mode will provide users an error experience when there are namespace conflicts to which a user will be able to resolve by defining one namespace per package source configuration.
+
+An error for this experience might look like:
+
+```console
+NUXXXX: Package namespace {0} is listed on the following sources: {1}. Only one unique package namespace can be defined across sources.
+```
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+#### Technical background - how NuGet package installation works today
+
+This approach builds on top of the current package installation behavior as described in the [package installation process](https://docs.microsoft.com/nuget/concepts/package-installation-process).
+
+A few other important concepts:
+
+- global packages folder - The installation directory for all packages in PackageReference. Think of this as the `Program Files` equivalent. All packages are consumed from this location. A single package installation can be shared safely among many different builds. A project *does not own* a specific package installation.
+The global packages folder is an *append only* resource. This means NuGet only ever installs a new package. There is **no** refreshing, or overriding, or `re-installing` packages of any kind.
+- When requesting an exact version, local, file sources are currently preferred - When installing a package, NuGet checks local sources independently before checking http sources.
```

Comment:
According to [Version ranges](https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#version-ranges), `1.0` is a minimum version, while `[1.0]` is an exact version. Does this mean that, if the project references `1.0`, then that is not an exact version and NuGet may choose version 1.0 from a remote package source even if version 1.0 is also available from a local package source? If so, that might be worth calling out here.

Reply:
Great point, what do you think @nkolev92 ?

Reply:
The explanation needs to rephrased.
Exact in this context really means, some project or package in the project graph, declared a dependency on that version as a min version. In other words, not floating. 

So `1.0` or `[1.0]` will both check local first. 

Does changing it to `specific` alleviate the confusion?

Reply:
No, "specific" would not be any clearer. I think this should list the combinations:

* `1.0` or `[1.0]` or `[1.0,3.0)` or `[1.0,3.0]` range, and 1.0 exists in a local source: Do not search remote sources.
* `1.0` or `[1.0,3.0)` or `[1.0,3.0]` range, and 1.1 exists in a local source but 1.0 doesn't: Check for 1.0 in local sources, then search local and remote sources in parallel.
* `(1.0,3.0)` or `(1.0,3.0]` range (exclusive minimum version): Search local and remote sources in parallel.
* `1.*` range (floating version): Search local and remote sources in parallel.

Then integrate those to the tables in [Package versioning](https://docs.microsoft.com/en-us/nuget/concepts/package-versioning) if the proposal is accepted.

Reply:
@KalleOlaviNiemitalo 

Please note that there are no behavior changes proposed here. 
This describes the current behavior. 
The preference for local sources when searching for a specific version for *any* reason was added as a response to customer feedback, and we're not proposing that to change in any way right now.

Reply:
We'd be happy to accept enhancements at the already documented doc https://docs.microsoft.com/en-us/nuget/concepts/package-installation-process. 

For now, I'd prefer not to elaborate on that in too much detail, as I don't want it to distract from the original intention of this design.